### DISCUSSION TITLE
Cloudflare Durable Objects - how to export classes for `durable_objects` bindings?

### DISCUSSION THREAD

**Discussion by jake-danton** [nitrojs/nitro#1832] [Q&A] [answered]
I am investigating getting nitro working with Cloudflare Durable Objects. In order to do so, you must define [bindings ](https://developers.cloudflare.com/durable-objects/get-started/#5-configure-durable-object-bindings) in `wrangler.toml` that maps a name in the Cloudflare `env` instance to an exported class. Example:

```
[durable_objects]
bindings = [
  { name = "someObject", class_name = "SomeObject" },
  { name = "someOtherObject", class_name = "SomeOtherObject" }
]
```

However, when building with the "cloudflare_module" preset, `index.mjs` only exports the [default object](https://developers.cloudflare.com/durable-objects/get-started/#4-instantiate-and-communicate-with-a-durable-object) with the required `fetch` method (`cloudflareModule`). Other classes marked as "export" in the src code are no longer exported in `index.mjs`.

How do I go about defining these "SomeObject" and "SomeOtherObject" [Durable Object classes](https://developers.cloudflare.com/durable-objects/get-started/#3-write-a-class-to-define-a-durable-object) such that they will be exported as well? Is there a specific directory? A decorator? Preset configuration?

Thanks!
Jake

**Comment by jake-danton** [comment]
@timhanlon  @pi0 I believe both of you have worked on this feature. If you point me in the right direction, I am happy to create a PR with more docs or sample code. Just don't see anything in https://github.com/unjs/nitro/blob/main/src/runtime/entries/cloudflare-module.ts to point me in a useful direction for adding more exports.

**Reply by timhanlon** [reply, marked as answer]
I dug this out of the Nuxt Discord, worth a shot:

https://discord.com/channels/473401852243869706/897487139888062506/1093592477073670244

https://github.com/nicgene/nuxt-durable-object

**Reply by jake-danton** [reply]
Awesome! I'll give this approach a try and update as needed. Thanks for the reference and speedy reply!

**Reply by Hebilicious** [reply]
For most cloudflare-y things (DO, queues, crons ...) with Nitro, you need to use this custom preset approach. The reason it is undocumented is because most users shouldn't rely on "hacky" workarounds, and nitro should provide first class support for these features eventually.

**Reply by jake-danton** [reply]
Good point. I really enjoy using Nitro, so I will go with the hack for now (it works for my case), but am super excited for full support for this and websockets!

Thanks to all the nitro contributors!

**Comment by jake-danton** [comment]
For anyone interested in how this looks, here is the simplest form I could get it to.

`./nitro.config.ts`
```
export default defineNitroConfig({
  preset: "./preset",
});
```

`./preset/nitro.config.ts`:
```
import { type NitroPreset } from "nitropack";
import { fileURLToPath } from "url";

export default <NitroPreset>{
  extends: "cloudflare-module",
  entry: fileURLToPath(new URL('./entry.ts', import.meta.url)),
};
```

`./preset/entry.ts`:
```
import entry from '#internal/nitro/entries/cloudflare-module'

export * from '../objects'
export default entry
```

`./objects/index.ts`
```
export { FirstDurableObject } from './FirstDurableObject'
export { SecondDurableObject } from './SecondDurableObject'
...
```

That way, all I have to do to add more durable objects is add a new file in `./objects` and add the export in `./objects/index.ts`.

And with the `export * from '../objects'` approach, I should be able to reuse this preset with any Nitro project I need to without additional configuration.

Really appreciate the Nitro team (and unjs in general!) for the flexible apis!

(Now to see if I can get Cloudflare Durable Objects Websockets working too 😁 )
